### PR TITLE
BUILD-87: secret configmap volume mounts in builds

### PR DESCRIPTION
**Pull request by coreydaley**
https://issues.redhat.com/browse/BUILD-87

### CODE REVIEW

File: pkg/build/builder/daemonless.go
Comment (adambkaplan) [inline pkg/build/builder/daemonless.go:337]: Why not accept an object that mirrors or extends the `BuildVolumeMount` object in the api? That way `append` can encode the data into the internal map, with the appropriate options added on.

```suggestion
type TransientMount struct {
   buildv1.BuildVolumeMount
   // additional buildah mount options not exposed to the Build API
  NoDev bool 
  NoExec bool
  NoSuid bool
}

func (t TransientMounts) append(mount TransientMount) error {
```

File: pkg/build/builder/daemonless.go
Comment (adambkaplan) [inline pkg/build/builder/daemonless.go:401]: If `append` accepts a fuller TransientMount object as an argument, then this logic is not repeated.

File: pkg/build/builder/daemonless.go
Comment (adambkaplan) [inline pkg/build/builder/daemonless.go:502]: Ditto - TransientMount object will abstract away the read-only bit.

**Review by adambkaplan — CHANGES_REQUESTED** [review]

File: pkg/build/builder/daemonless.go
Comment (adambkaplan) [inline pkg/build/builder/daemonless.go:348]: Need an else for read-only
```suggestion
	if !tm.BuildVolumeMount.ReadOnly {
		mountOptions = append(mountOptions, "rw")
	} else {
		mountOptions = append(mountOptions, "ro") 
	}
```

File: pkg/build/builder/daemonless.go
Comment (adambkaplan) [inline pkg/build/builder/daemonless.go:423]: Can get rid of this

File: pkg/build/builder/daemonless.go
Comment (adambkaplan) [inline pkg/build/builder/daemonless.go:427]: Now I see what you are doing in openshift-controller-manager - you are laying the mounts out based on their destination paths. It helps alleviate the volume name collision problem. `filepath.Join` also strips the leading `/` if we are given an absolute file path, which is good. We should probably convert relative file paths to an absolute filepath within buildah's runtime environment.

File: pkg/build/builder/daemonless.go
Comment (adambkaplan) [inline pkg/build/builder/daemonless.go:358]: See my note about `MountPath` being relative - we should convert this to an absolute path.

**Comment by gabemontero** [comment]
FYI there is auto linking of PRs to Jira items if you 

/retitle BUILD-87 secret configmap volume mounts in builds

**Comment by gabemontero** [comment]
/retitle BUILD-87: secret configmap volume mounts in builds

**Comment by gabemontero** [comment]
Course I don't have permissions to do it @coreydaley :-)

**Comment by adambkaplan** [comment]
/hold

Given the complexity of the feature and the time remaining before 4.8 reaches "final freeze" period, we are going to hold this until master opens for 4.9

**Comment by coreydaley** [comment]
@adambkaplan ptal

**Comment by xiuwang** [comment]
Test on PR launched cluster launch openshift/builder#245,openshift/openshift-controller-manager#183,openshift/openshift-apiserver#208,openshift/oc#843 following test cases https://url.corp.redhat.com/buildvolumes , no critical issues found.

/label qe-approved

**Comment by coreydaley** [comment]
@adambkaplan @gabemontero ptal, tests are passing and it is ready for review.

File: pkg/build/builder/transient_mounts.go
Comment (gabemontero) [inline pkg/build/builder/transient_mounts.go:60]: might be the first time I've seen in practice a non-struct ptr used for this .... cool :-)

File: pkg/build/builder/transient_mounts.go
Comment (gabemontero) [inline pkg/build/builder/transient_mounts.go:25]: don't we typically employ `tm *TransientMount)` here @coreydaley ?

if there is a specific reason why you did not do it that way, I'm missing it (and perhaps a clarifying comment wrt that is needed)

File: pkg/build/builder/transient_mounts.go
Comment (gabemontero) [reply pkg/build/builder/transient_mounts.go:60]: btw, the map ref is a ptr under the cover iirc correctly, hence you need nil checks when checking map variables, re: related to my comment above with the funcs off of the `TransientMount` struct

File: pkg/build/builder/util.go
Comment (gabemontero) [inline pkg/build/builder/util.go:158]: Did I see an exchange in a library-go PR that landed with *NOT* including a method like this there?

I ask because I presume (though I have not visited your PR there super recently) there is a need for such a method in openshift-controller-manager, and we'd ideally share the impl (albeit this is a simple thing).

But it may not be a given, and if so, say as much @coreydaley and we can resolve this thread ... but I'll mention it here anyway just in case.

So just let me know and we can go from there.

Comment (gabemontero) [reply pkg/build/builder/util.go:158]: we do have https://github.com/openshift/runtime-utils already for sharing build specific logic that could be a landing spot for such common methods that is specific to builds vs. library-go

File: pkg/build/builder/util.go
Comment (gabemontero) [inline pkg/build/builder/util.go:164]: same thing here wrt shared functions between builder and openshift-controller-manager

**Review by gabemontero — COMMENTED** [review]
A few comments, but no big blockers ... other than perhaps sharing some functions between builder and OCM in https://github.com/openshift/runtime-utils most likely these are just questions of mine you can answer @coreydaley and we can move on.

File: pkg/build/builder/util.go
Comment (coreydaley) [reply pkg/build/builder/util.go:158]: It was decided by the powers that be that this code should be duplicated between the repos instead of shared via library-go, and as such, the library-go pull request has been closed.

As for `runtime-utils`, all I see there is a `doc.go` in the build directory?

File: pkg/build/builder/util.go
Comment (coreydaley) [reply pkg/build/builder/util.go:164]: It was decided by the powers that be that this code should be duplicated between the repos instead of shared via library-go, and as such, the library-go pull request has been closed.

File: pkg/build/builder/transient_mounts.go
Comment (coreydaley) [reply pkg/build/builder/transient_mounts.go:25]: Since this method is not mutating the underlying data structure the pointer type is not needed. I'm not sure that a comment for that is needed.

https://gobyexample.com/methods

File: pkg/build/builder/transient_mounts.go
Comment (coreydaley) [reply pkg/build/builder/transient_mounts.go:60]: > might be the first time I've seen in practice a non-struct ptr used for this .... cool :-)

I'm not going to say that this is common practice, or even the correct thing to do, but it worked :)

File: pkg/build/builder/util.go
Comment (gabemontero) [reply pkg/build/builder/util.go:158]: > It was decided by the powers that be that this code should be duplicated between the repos instead of shared via library-go, and as such, the library-go pull request has been closed.
> 

ok so my recollection was correct in this case

> As for `runtime-utils`, all I see there is a `doc.go` in the build directory?

no there is also a pkg directory, with currently a "registries" subdir

technically speaking, per the readme, this repo was a common ground for https://github.com/containers related shared libs that build code would leverage, but there is no reason IMO would could not expand that a bit

@adambkaplan and @bparees are in the OWNER file

also note, that repo arose when similar utility functions were rejected by library-go

File: pkg/build/builder/transient_mounts.go
Comment (coreydaley) [reply pkg/build/builder/transient_mounts.go:60]: > btw, the map ref is a ptr under the cover iirc correctly, hence you need nil checks when checking map variables, re: related to my comment above with the funcs off of the `TransientMount` struct

I believe that `ok := t[mount.Destination];` equates to a `nil` check, as it returns `false` if the key does not exist. Beyond that we are really only doing a few operations here:
 - Checking for duplicate keys
 - Adding a key/value pair
 - Outputting the map as a slice of strings constructed from the transient mount that is the value
If there is a specific place you think that I should be checking for `nil` let me know and I will add it.

Comment (gabemontero) [reply pkg/build/builder/transient_mounts.go:60]: OK, then I'll make an official "nit" req to change it to a pointer ref

**Review by adambkaplan — APPROVED** [review]
/approve

@rolfedh @RickJWagner ptal - we need "docs-approved" and "px-approved" labels before we merge.

File: pkg/build/builder/transient_mounts.go
Comment (gabemontero) [reply pkg/build/builder/transient_mounts.go:25]: I don't think this one should be marked resolved yet @coreydaley

File: pkg/build/builder/transient_mounts.go
Comment (gabemontero) [reply pkg/build/builder/transient_mounts.go:60]: yeah I'm good with this one ... go ahead and mark this one resolved ... but the other thread needs to be "unresolved"

File: pkg/build/builder/transient_mounts.go
Comment (coreydaley) [reply pkg/build/builder/transient_mounts.go]: updated in current push

**Review by gabemontero — COMMENTED** [review]
ok, employing the same 

/lgtm

with 

/hold

while we wait on @adambkaplan to chime in on expanding https://github.com/openshift/runtime-utils to capture these common build specific constants and funcs that are denied inclusion into openshift/library-go

if we don't use it, unhold @coreydaley

**Comment by adambkaplan** [comment]
Keeping the hold for docs/px approvals.

I don't think we should expand the scope of runtime-utils to add this copied code, since it pollutes that repo's purpose. I am disappointed that our PR to library-go was rejected, since other code in that repository doesn't meet their criteria.

**Pull request merged**
